sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,328 @@ def transform(self, X):
+        X = check_array(X, accept_sparse=False, dtype=np.float64,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        self.weights = _check_weights(self.weights)
+
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed.")
+
+        # Check if % missing in any column > col_max_missing
+        mask = _get_mask(X, self.missing_values)
+        if np.any(mask.sum(axis=0) > (X.shape[0] * self.col_max_missing)):
+            raise ValueError("Some column(s) have more than {}% missing values"
+                             .format(self.col_max_missing*100))
+        X_col_means = np.ma.array(X, mask=mask).mean(axis=0).data
+
+        # Check if % missing in any row > row_max_missing
+        bad_rows = mask.sum(axis=1) > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. These "
+                "rows are not included as donor neighbors."
+                    .format(self.row_max_missing * 100))
+
+            # Remove rows that have more than row_max_missing % missing
+            X = X[~bad_rows, :]
+
+        # Check if sufficient neighboring samples available
+        if X.shape[0] < self.n_neighbors:
+            raise ValueError("There are only %d samples, but n_neighbors=%d."
+                             % (X.shape[0], self.n_neighbors))
+
+        # Instantiate NN object, get column means, and store in statistics_
+        neigh = NearestNeighbors(n_neighbors=self.n_neighbors,
+                                 metric=self.metric,
+                                 metric_params={"missing_values":
+                                                self.missing_values})
+        self._fitted_neighbors = neigh.fit(X)
+        self.statistics_ = X_col_means
+
+        return self
+
+    def _transform(self, X, adjusted_n_neighbors):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        adjusted_n_neighbors : int
+            Indicates whether to pass n_neighbors or n_neighbors+1 to
+            _tranform().
+            Calling transform() automatically sets this to self.n_neighbors
+            while fit_transform() sets it to self.n_neighbors + 1.
+        """
+        # Import(s) here to avoid circular import
+        from ..neighbors.base import _get_weights
+
+        check_is_fitted(self, 'statistics_')
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
```

Comment:
It would be good if we could avoid validating so often (especially this relatively expensive O(n) validation, unlike O(1) shape/format validation)... atm in fit_transform, we seem to be validating in:
* `KNNImputer.fit`
* `KNNImputer._transform`
* `NearestNeighbors.fit`
* `NearestNeighbors.kneighbors`
* `pairwise_distances` checks the row-entirely-missing case
* `masked_euclidean_distances`

Reply:
Yeah makes sense. Which of these would you preferred to be removed?

Reply:
FWIW now knnimputer._transform does not does not check if X is passed with fit_transform. And in NearestNeighbors it appears that kneighbors() does not check if query_is_train. Not sure what you mean by pairwise checks row_entirely_missing (these should be rejected at the outset by knnimputer? unless you mean generally outside of imputer). It appears that a "typical" imputation call using fit_transform would have to pass through three checks right now (from list above): (1), (3), and (6)?

Reply:
One option would be to pass around keywords indicating X has been checked, but not sure if bloating the api might be worth it